[threaded-animations] creating a new animation removes the timeline for a previous animation

graouts opened the pull request:
#### f27e060bbaf0b9b69e085ab05fbe447cc17536d0
<pre>
[threaded-animations] creating a new animation removes the timeline for a previous animation
<a href="https://bugs.webkit.org/show_bug.cgi?id=303655">https://bugs.webkit.org/show_bug.cgi?id=303655</a>
<a href="https://rdar.apple.com/165937022">rdar://165937022</a>

Reviewed by Simon Fraser.

If a scroll-driven animation is created for a given target, and later on a different
scroll-driven animation is created for another target, then the timeline for the first
animation is removed from the remote layer tree.

This is due to our ill-conceived method to gather relevant timelines for a given remote
layer tree transaction and updating them upon reception in the remote layer tree. Indeed,
until now, when we would create an update, we would gather all elements marked as required
an update in `AcceleratedEffectStackUpdater::update()` and keep a list of all timelines
seen for animations targeting said elements. For each such update, that list of timelines
would replace the one from the previous update, and would be made available in the remote
layer tree transaction.

As such, if we weren&apos;t updating animations for the same elements that were previously updated,
their timelines would be forgotten and the logic on the remote layer tree side would assume
they needed to be removed.

Additionally, we also naively always considered that timelines were part of the remote layer
tree transaction, assuming an empty list of timelines meant all timelines needed to be removed.
This meant that, even on pages where we never had timelines, we would attempt to update our
timelines registration on each remote layer tree transaction.

We overhaul this entire system by doing two main things.

First, we rethink the accelerated timeline ownership in the Web process such that the only
object that has a strong pointer (`RefPtr`) are the accelerated effects that are stored on
the GraphicsLayer matching the animation&apos;s target element. The `AcceleratedEffectStackUpdater`
maintains a list of known timelines but keeps weak pointers (`WeakPtr`), so does `AnimationTimeline`
for its accelerated representation. This now means that if, during an update, all of the effects
linked to a given timeline are removed, the timeline will no longer have strong references
and `AcceleratedEffectStackUpdater` will be able to check for null pointers and determine these
timelines indeed need to be destroyed.

Second, we now have a dedicated `TimelinesUpdate` struct used in remote layer tree transactions
sorting timelines in three categories: `created`, `modified` and `destroyed`. In the new system
outlined above, `AcceleratedEffectStackUpdater` is able to determine, during an update, how to
categorize timelines, making it clear on the receiving end, the remote layer tree, which timelines
to register, update and remove. Additionally, we can now determine if the remote layer tree
transaction contains any timelines to update and will do no processing whatsoever if we figure
out there are no timelines in any of those three buckets.

A new test validates that this new approach works as expected and we can update animations associated
with different targets across different updates.

Additional details as necessary in the function-per-function ChangeLog below.

Test: webanimations/threaded-animations/remote-progress-based-timeline-preservation-with-new-remote-animation-stack.html

* LayoutTests/webanimations/threaded-animations/remote-progress-based-timeline-preservation-with-new-remote-animation-stack-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/remote-progress-based-timeline-preservation-with-new-remote-animation-stack.html: Added.
* Source/WebCore/animation/AcceleratedEffectStackUpdater.cpp:
(WebCore::AcceleratedEffectStackUpdater::update):
(WebCore::AcceleratedEffectStackUpdater::scrollTimelineDidChange):
(WebCore::AcceleratedEffectStackUpdater::takeTimelinesUpdate):
* Source/WebCore/animation/AcceleratedEffectStackUpdater.h:
(WebCore::AcceleratedEffectStackUpdater::timelines const): Deleted.
* Source/WebCore/animation/AnimationTimeline.cpp:
(WebCore::AnimationTimeline::acceleratedRepresentation): Return a Ref that will be stored by the `AcceleratedEffect` caller, keeping only a
weak pointer to that timeline as a member.
(WebCore::AnimationTimeline::runPostRenderingUpdateTasks): Make sure _not_ to clear the accelerated representation pointer as we now want
to keep the `AcceleratedTimeline` instance as long as it is used by an `AcceleratedEffect`.
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::acceleratedTimelineIdentifier const):
(WebCore::AnimationTimeline::acceleratedTimelineIdentifierForTesting const): Deleted.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::acceleratedRepresentation): Return a `Ref` instead of a `RefPtr` now that we know `AcceleratedEffect` will also
return a `Ref` always.
(WebCore::KeyframeEffect::animationProgressBasedTimelineSourceDidChangeMetrics): No need to trigger an animation update here, scroll timelines
themselves will inform `AcceleratedEffectStackUpdater` that they require an update, not needing the animation stack to be updated.
(WebCore::KeyframeEffect::updatedAcceleratedRepresentation): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::cacheCurrentTime):
(WebCore::ScrollTimeline::sourceMetricsDidChange): Introduce this new protected method to be called also by `ViewTimeline` when metrics change,
informing `AcceleratedEffectStackUpdater` that the accelerated representation needs to be updated.
(WebCore::ScrollTimeline::scheduleAcceleratedRepresentationUpdate):
(WebCore::ScrollTimeline::computeProgressResolutionData const): New function to compute the `ProgressResolutionData` such that it can be used
both when an update is needed, or when the accelerated representation is created.
(WebCore::ScrollTimeline::updateAcceleratedRepresentation):
(WebCore::ScrollTimeline::createAcceleratedRepresentation const):
(WebCore::ScrollTimeline::scrollingNodeIDForTesting const):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::cacheCurrentTime):
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
(WebCore::AcceleratedEffect::create): Switch this method to a `Ref` which is a more common idiom that returning a `RefPtr`. The logic previously
used to determine whether this effect should be accelerated is now contained in `RenderLayerBacking::updateAcceleratedEffectsAndBaseValues()`.
(WebCore::AcceleratedEffect::AcceleratedEffect): Create and store the accelerated timeline as a strong pointer.
* Source/WebCore/platform/animation/AcceleratedEffect.h:
* Source/WebCore/platform/animation/AcceleratedTimeline.cpp:
(WebCore::AcceleratedTimeline::setProgressResolutionData):
* Source/WebCore/platform/animation/AcceleratedTimeline.h:
(WebCore::AcceleratedTimelinesUpdate::isEmpty const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::identifierForTimeline const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::updateTimelinesRegistration):
(WebKit::RemoteLayerTreeDrawingAreaProxy::animationStackForNodeWithIDForTesting const):
(WebKit::RemoteLayerTreeDrawingAreaProxy::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree): Only initiate a timelines registration update if we have any data.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimelineRegistry.cpp:
(WebKit::RemoteMonotonicTimelineRegistry::update): change the registration code to work the new `TimelinesUpdate` struct and its
three `created`, `modified` and `destroyed` buckets.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimelineRegistry.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.cpp:
(WebKit::RemoteProgressBasedTimelineRegistry::update): change the registration code to work the new `TimelinesUpdate` struct and
its three `created`, `modified` and `destroyed` buckets.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::updateTimelinesRegistration):
(WebKit::RemoteScrollingCoordinatorProxy::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::updateTimelinesRegistration):
(WebKit::RemoteScrollingTree::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateTimelinesRegistration):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.mm:
(WebKit::RemoteLayerTreeEventDispatcher::updateTimelinesRegistration):
(WebKit::RemoteLayerTreeEventDispatcher::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::updateTimelinesRegistration):
(WebKit::RemoteScrollingCoordinatorProxyMac::updateTimelineRegistration): Deleted.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::willCommitLayerTree): Assemble all the `TimelinesUpdate` for all documents into a single update.

Canonical link: <a href="https://commits.webkit.org/304042@main">https://commits.webkit.org/304042@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48d5bb5dc7efc7f116888bd20ad2371f5005914c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134347 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45563 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141894 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86351 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136217 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7415 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6689 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102699 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1cbe818f-ca5d-4482-b533-a2da459e2c34) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137294 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5166 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120402 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83490 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e83a2609-6a30-4ba3-9681-80647e576127) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5033 "Found 1 new test failure: imported/w3c/web-platform-tests/fetch/http-cache/cc-request.any.worker.html (failure)") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2652 "Passed tests") | [⏳ 🛠 wpe-cairo-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-Cairo-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114237 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38543 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144571 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6502 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39126 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111098 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6580 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5462 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111365 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4869 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116678 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60285 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20751 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6554 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34878 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6382 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70117 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6617 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6494 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/AcceleratedEffectStackUpdater.cpp:
```diff
@@ -58,7 +60,15 @@ void AcceleratedEffectStackUpdater::update()
+
+    for (auto& timeline : timelinesInUpdate) {
+        auto& timelineIdentifier = timeline->identifier();
+        if (!m_timelines.contains(timelineIdentifier)) {
```
`contains` followed by `set` is two hash lookups. Consider `add` and checking `addResult`.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -420,19 +438,32 @@ Ref<AcceleratedTimeline> ScrollTimeline::createAcceleratedRepresentation() const
 
-    return AcceleratedTimeline::create(m_acceleratedTimelineIdentifier, {
+    return {
         *sourceScrollableArea->scrollingNodeID(),
```
I like the `{ .foo = bar, .baz = ... }` format that shows the field names.

graouts replied:
That's _much_ better, will make that change.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimelineRegistry.cpp:
```diff
@@ -36,33 +36,58 @@ namespace WebKit {
-        m_timelines.remove(processIdentifier);
-        return;
-    }
+#if !defined(NDEBUG)
```
`#if ASSERT_ENABLED`

graouts committed:
[threaded-animations] creating a new animation removes the timeline for a previous animation
https://bugs.webkit.org/show_bug.cgi?id=303655
rdar://165937022

Reviewed by Simon Fraser.

If a scroll-driven animation is created for a given target, and later on a different
scroll-driven animation is created for another target, then the timeline for the first
animation is removed from the remote layer tree.

This is due to our ill-conceived method to gather relevant timelines for a given remote
layer tree transaction and updating them upon reception in the remote layer tree. Indeed,
until now, when we would create an update, we would gather all elements marked as required
an update in `AcceleratedEffectStackUpdater::update()` and keep a list of all timelines
seen for animations targeting said elements. For each such update, that list of timelines
would replace the one from the previous update, and would be made available in the remote
layer tree transaction.

As such, if we weren't updating animations for the same elements that were previously updated,
their timelines would be forgotten and the logic on the remote layer tree side would assume
they needed to be removed.

Additionally, we also naively always considered that timelines were part of the remote layer
tree transaction, assuming an empty list of timelines meant all timelines needed to be removed.
This meant that, even on pages where we never had timelines, we would attempt to update our
timelines registration on each remote layer tree transaction.

We overhaul this entire system by doing two main things.

First, we rethink the accelerated timeline ownership in the Web process such that the only
object that has a strong pointer (`RefPtr`) are the accelerated effects that are stored on
the GraphicsLayer matching the animation's target element. The `AcceleratedEffectStackUpdater`
maintains a list of known timelines but keeps weak pointers (`WeakPtr`), so does `AnimationTimeline`
for its accelerated representation. This now means that if, during an update, all of the effects
linked to a given timeline are removed, the timeline will no longer have strong references
and `AcceleratedEffectStackUpdater` will be able to check for null pointers and determine these
timelines indeed need to be destroyed.

Second, we now have a dedicated `TimelinesUpdate` struct used in remote layer tree transactions
sorting timelines in three categories: `created`, `modified` and `destroyed`. In the new system
outlined above, `AcceleratedEffectStackUpdater` is able to determine, during an update, how to
categorize timelines, making it clear on the receiving end, the remote layer tree, which timelines
to register, update and remove. Additionally, we can now determine if the remote layer tree
transaction contains any timelines to update and will do no processing whatsoever if we figure
out there are no timelines in any of those three buckets.

A new test validates that this new approach works as expected and we can update animations associated
with different targets across different updates.

Additional details as necessary in the function-per-function ChangeLog below.

Test: webanimations/threaded-animations/remote-progress-based-timeline-preservation-with-new-remote-animation-stack.html

* LayoutTests/webanimations/threaded-animations/remote-progress-based-timeline-preservation-with-new-remote-animation-stack-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/remote-progress-based-timeline-preservation-with-new-remote-animation-stack.html: Added.
* Source/WebCore/animation/AcceleratedEffectStackUpdater.cpp:
(WebCore::AcceleratedEffectStackUpdater::update):
(WebCore::AcceleratedEffectStackUpdater::scrollTimelineDidChange):
(WebCore::AcceleratedEffectStackUpdater::takeTimelinesUpdate):
* Source/WebCore/animation/AcceleratedEffectStackUpdater.h:
(WebCore::AcceleratedEffectStackUpdater::timelines const): Deleted.
* Source/WebCore/animation/AnimationTimeline.cpp:
(WebCore::AnimationTimeline::acceleratedRepresentation): Return a Ref that will be stored by the `AcceleratedEffect` caller, keeping only a
weak pointer to that timeline as a member.
(WebCore::AnimationTimeline::runPostRenderingUpdateTasks): Make sure _not_ to clear the accelerated representation pointer as we now want
to keep the `AcceleratedTimeline` instance as long as it is used by an `AcceleratedEffect`.
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::acceleratedTimelineIdentifier const):
(WebCore::AnimationTimeline::acceleratedTimelineIdentifierForTesting const): Deleted.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::acceleratedRepresentation): Return a `Ref` instead of a `RefPtr` now that we know `AcceleratedEffect` will also
return a `Ref` always.
(WebCore::KeyframeEffect::animationProgressBasedTimelineSourceDidChangeMetrics): No need to trigger an animation update here, scroll timelines
themselves will inform `AcceleratedEffectStackUpdater` that they require an update, not needing the animation stack to be updated.
(WebCore::KeyframeEffect::updatedAcceleratedRepresentation): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::cacheCurrentTime):
(WebCore::ScrollTimeline::sourceMetricsDidChange): Introduce this new protected method to be called also by `ViewTimeline` when metrics change,
informing `AcceleratedEffectStackUpdater` that the accelerated representation needs to be updated.
(WebCore::ScrollTimeline::scheduleAcceleratedRepresentationUpdate):
(WebCore::ScrollTimeline::computeProgressResolutionData const): New function to compute the `ProgressResolutionData` such that it can be used
both when an update is needed, or when the accelerated representation is created.
(WebCore::ScrollTimeline::updateAcceleratedRepresentation):
(WebCore::ScrollTimeline::createAcceleratedRepresentation const):
(WebCore::ScrollTimeline::scrollingNodeIDForTesting const):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::cacheCurrentTime):
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
(WebCore::AcceleratedEffect::create): Switch this method to a `Ref` which is a more common idiom that returning a `RefPtr`. The logic previously
used to determine whether this effect should be accelerated is now contained in `RenderLayerBacking::updateAcceleratedEffectsAndBaseValues()`.
(WebCore::AcceleratedEffect::AcceleratedEffect): Create and store the accelerated timeline as a strong pointer.
* Source/WebCore/platform/animation/AcceleratedEffect.h:
* Source/WebCore/platform/animation/AcceleratedTimeline.cpp:
(WebCore::AcceleratedTimeline::setProgressResolutionData):
* Source/WebCore/platform/animation/AcceleratedTimeline.h:
(WebCore::AcceleratedTimelinesUpdate::isEmpty const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::identifierForTimeline const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::updateTimelinesRegistration):
(WebKit::RemoteLayerTreeDrawingAreaProxy::animationStackForNodeWithIDForTesting const):
(WebKit::RemoteLayerTreeDrawingAreaProxy::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree): Only initiate a timelines registration update if we have any data.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimelineRegistry.cpp:
(WebKit::RemoteMonotonicTimelineRegistry::update): change the registration code to work the new `TimelinesUpdate` struct and its
three `created`, `modified` and `destroyed` buckets.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimelineRegistry.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.cpp:
(WebKit::RemoteProgressBasedTimelineRegistry::update): change the registration code to work the new `TimelinesUpdate` struct and
its three `created`, `modified` and `destroyed` buckets.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::updateTimelinesRegistration):
(WebKit::RemoteScrollingCoordinatorProxy::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::updateTimelinesRegistration):
(WebKit::RemoteScrollingTree::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateTimelinesRegistration):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.mm:
(WebKit::RemoteLayerTreeEventDispatcher::updateTimelinesRegistration):
(WebKit::RemoteLayerTreeEventDispatcher::updateTimelineRegistration): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::updateTimelinesRegistration):
(WebKit::RemoteScrollingCoordinatorProxyMac::updateTimelineRegistration): Deleted.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::willCommitLayerTree): Assemble all the `TimelinesUpdate` for all documents into a single update.

Canonical link: https://commits.webkit.org/304042@main
